### PR TITLE
lsp: text_document.version may be vim.NIL not nil

### PR DESCRIPTION
**Pull request by h-michael**
I take over #12191

cc: @ckipp01

### CODE REVIEW

File: test/functional/plugin/lsp_spec.lua [hunk 1]
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
...
 
+  describe('apply_text_document_edit', function()
+    local target_bufnr
+    before_each(function()
+      target_bufnr = exec_lua [[
+        local bufnr = vim.fn.bufadd("fake/uri")
+        local lines = {"1st line of text", "2nd line of text"}
+        vim.api.nvim_buf_set_lines(bufnr, 0, 1, false, lines)
+        return bufnr
+      ]]
```
Comment (ckipp01) [inline test/functional/plugin/lsp_spec.lua:828]: Ahh, this is super helpful to see how to do. 👍

File: test/functional/plugin/lsp_spec.lua [hunk 2]
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
     end)
   end)
 
+  describe('apply_text_document_edit', function()
+    local target_bufnr
+    before_each(function()
+      target_bufnr = exec_lua [[
+        local bufnr = vim.fn.bufadd("fake/uri")
+        local lines = {"1st line of text", "2nd line of text"}
```
Comment (clason) [inline test/functional/plugin/lsp_spec.lua:825]: Please add a non-ASCII character to one of the lines (maybe "Ä 2nd line of text"?) to test for encoding.

File: test/functional/plugin/lsp_spec.lua [hunk 3]
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
...
+        local target_bufnr = args[2]
+        vim.lsp.util.buf_versions[target_bufnr] = 4
+        vim.lsp.util.apply_text_document_edit(...)
+      ]], text_document_edit, target_bufnr)
+      eq({
+        'First line of text';
+        '2nd line of text';
+      }, buf_lines(target_bufnr))
+    end)
+    it('correctly goes ahead with the edit whe the version is vim.NIL', function()
```
Comment (clason) [inline test/functional/plugin/lsp_spec.lua:851]: typo: "when" (or "correctly applies the edit if the version...")

File: test/functional/plugin/lsp_spec.lua (hunk 2 above)
Comment (teto) [reply test/functional/plugin/lsp_spec.lua:825]: don't. Let's keep meaningful commits, non-ASCII characters should be merged with the same commits that introduce supports for them i.e., https://github.com/neovim/neovim/pull/12223

Comment (clason) [reply test/functional/plugin/lsp_spec.lua:825]: No, that commit didn't introduce support, non-ASCII characters should have worked from the start, so it fixes a bug this test would have caught.

Comment (clason) [reply test/functional/plugin/lsp_spec.lua:825]: (But once the test is merged, I can try to extend it with non-ASCII characters. I'd prefer not to merge this PR into my PR, though.)

**Comment by h-michael** [comment]
@teto @clason 
I've fixed a bit and add a non-ASCII characters test case pending.

**Pull request merged**